widget/embedded/src/components/ModalStatefulConnect/ModalStatefulConnect.tsx

```diff
@@ -0,0 +1,142 @@
+} from '../../hooks/useStatefulConnect';
+import { useWalletList } from '../../hooks/useWalletList';
+import { WatermarkedModal } from '../common/WatermarkedModal';
+import {
+  ConnectStatus,
+  DerivationPath,
+  Namespaces,
+} from '../WalletStatefulConnect';
+
+interface PropTypes {
+  wallet: WalletInfoWithExtra | undefined;
+  onClose: () => void;
+  // When `handleConnect` runs **successfully**, this will be called afterwards.
+  onConnect?: (result: Result) => void;
+}
+
+export function ModalStatefulConnect(props: PropTypes) {
+  const [isConnecting, setIsConnecting] = useState(false);
+  const [error, setError] = useState<string>();
+  const catchErrorOnHandle = (error: Error) => {
+    setError(error.message);
+  };
+  const { terminateConnectingWallets } = useWalletList();
+
+  const {
+    handleConnect,
+    handleDerivationPath,
+    handleNamespace,
+    getState,
+    resetState,
+  } = useStatefulConnect();
+
+  const handleConfirmNamespaces = (selectedNamespaces: Namespace[]) => {
+    handleNamespace(props.wallet!, selectedNamespaces).catch(
+      catchErrorOnHandle
+    );
+  };
+
+  const handleDerivationPathConfirm = (derivationPath: string) => {
+    if (!derivationPath) {
+      throw new Error(
+        "Derivation path is empty. Please make sure you've filled the field correctly."
+      );
+    }
+
+    handleDerivationPath(derivationPath).catch(catchErrorOnHandle);
+  };
+
+  const handleClosingModal = () => {
+    setIsConnecting(false);
+  };
+
+  const resetModalState = () => {
+    setError(undefined);
+    resetState();
+    setIsConnecting(false);
+  };
+
+  useEffect(() => {
+    if (props.wallet) {
+      resetModalState();
+
+      const beforeConnecting = () => {
+        setIsConnecting(true);
+      };
+
+      const afterConnected = (result: Result) => {
+        if (
+          [
+            ResultStatus.Connected,
+            ResultStatus.Disconnected,
+            ResultStatus.DisconnectedUnhandled,
+          ].includes(result.status)
+        ) {
+          handleClosingModal();
+        }
+
+        if (props.onConnect) {
+          props.onConnect(result);
+        }
+      };
+
+      beforeConnecting();
+      handleConnect(props.wallet, {
+        disconnectIfConnected: true,
+      })
+        .then(afterConnected)
+        .catch(catchErrorOnHandle);
+    }
+  }, [props.wallet]);
+
+  const isOnStatus = getState().status === 'init' && !!props.wallet;
+  const isOnNamespace = getState().status === 'namespace';
+  const isOnDerivationPath = getState().status === 'derivationPath';
+
+  return (
+    <WatermarkedModal
+      open={isConnecting}
+      onClose={handleClosingModal}
+      onExit={() => {
+        resetModalState();
+        terminateConnectingWallets();
+
+        if (props.onClose) {
+          props.onClose();
+        }
+      }}
+      container={
+        document.getElementById(WIDGET_UI_ID.SWAP_BOX_ID) || document.body
+      }>
+      {isOnStatus ? (
```

Comment:
I think using a function to render the content of modal is better than this complex conditional rendering.

Reply:
I think of something like this:

```

  const renderModalContent = () => {
    if (isOnStatus(getState, props.wallet)) {
      return (
        <>
          <ConnectStatus wallet={props.wallet} error={error} />
          <Divider direction="vertical" size={32} />
        </>
      );
    } else if (isOnNamespace(getState)) {
      return (
        <Namespaces
          onConfirm={handleConfirmNamespaces}
          value={getState().namespace}
        />
      );
    }
    return (
      <DerivationPath
        onConfirm={handleDerivationPathConfirm}
        value={getState().derivationPath}
      />
    );
  };

  return (
    <WatermarkedModal
      open={isConnecting}
      onClose={handleClosingModal}
      onExit={() => {
        resetModalState();
        terminateConnectingWallets();

        if (props.onClose) {
          props.onClose();
        }
      }}
      container={getContainer()}>
      {renderModalContent()}
    </WatermarkedModal>
  );
  ```

Reply:
It doesn't add any value (like readability or improvement) and seems same as the original code to me. So I prefer to not change.